### PR TITLE
feat: add isProfilePublic setting for accessing user info

### PR DESCRIPTION
**Pull request by seriouszyx**
Fix: #581 

add a organization option `enableLimitedAccess`

image

before enabled

image

after enabled

image

Signed-off-by: Yixiang Zhao <seriouszyx@foxmail.com>

### CODE REVIEW

File: controllers/user.go [hunk 1]
```diff
@@ -87,6 +87,16 @@ func (c *ApiController) GetUser() {
 	id := c.Input().Get("id")
 	owner := c.Input().Get("owner")
 	email := c.Input().Get("email")
+	organization := object.GetOrganization(fmt.Sprintf("%s/%s", "admin", strings.Split(id, "/")[0]))
+
+	if organization.EnableLimitedAccess {
```
Comment (hsluoyz) [inline controllers/user.go:92]: add `IsProfilePublic` boolean value, true by default

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: object/check.go [hunk 2]
```diff
@@ -195,3 +196,32 @@ func CheckUserPassword(organization string, username string, password string) (*
...
+	} else {
+		requestUser := GetUser(requestUserId)
+		if requestUser == nil {
+			return false, fmt.Errorf("session outdated, please login again")
+		}
+		if requestUser.IsGlobalAdmin {
+			hasPermission = true
+		} else if requestUserId == userId {
+			hasPermission = true
+		} else if targetUser.Owner == requestUser.Owner && requestUser.IsAdmin {
```
Comment (hsluoyz) [inline object/check.go:222]: the normal users from the same org should be able to see other's profile.

File: web/src/OrganizationListPage.js [hunk 3]
```diff
@@ -39,6 +39,7 @@ class OrganizationListPage extends BaseListPage {
       tags: [],
       masterPassword: "",
       enableSoftDeletion: false,
+      enableLimitedAccess: false,
     }
   }
```
Comment (hsluoyz) [inline web/src/OrganizationListPage.js:44]: Need to add frontend 404 page for other's profile.

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: controllers/user.go [hunk 4]
```diff
@@ -87,6 +87,16 @@ func (c *ApiController) GetUser() {
 	id := c.Input().Get("id")
 	owner := c.Input().Get("owner")
 	email := c.Input().Get("email")
+	organization := object.GetOrganization(fmt.Sprintf("%s/%s", "admin", strings.Split(id, "/")[0]))
```
Comment (hsluoyz) [inline controllers/user.go:90]: use: GetOwnerAndNameFromId

File: web/src/UserEditPage.js [hunk 5]
```diff
@@ -59,10 +59,14 @@ class UserEditPage extends React.Component {
 
   getUser() {
     UserBackend.getUser(this.state.organizationName, this.state.userName)
-      .then((user) => {
-        this.setState({
-          user: user,
-        });
+      .then((data) => {
+        if (data.status === "error") {
+          window.location.href = "/404";
```
Comment (hsluoyz) [inline web/src/UserEditPage.js:64]: don't redirect

File: object/check.go [hunk 6]
```diff
@@ -195,3 +196,36 @@ func CheckUserPassword(organization string, username string, password string) (*
...
+		} else if targetUser.Owner == requestUser.Owner {
+			if strict {
+				hasPermission = requestUser.IsAdmin
+			} else {
+				hasPermission = true
+			}
+		}
+	}
+	return hasPermission, fmt.Errorf("you don't have the permission to do this")
+}
```
Comment (hsluoyz) [inline object/check.go:231]: new line

File: controllers/user.go [hunk 7]
```diff
@@ -87,6 +87,16 @@ func (c *ApiController) GetUser() {
...
+
+	if organization.IsProfilePublic {
```
Comment (hsluoyz) [inline controllers/user.go:92]: wrong logic?

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Comment by seriouszyx** [comment]
access user in the same org

image

access user in the different org

image

**Review by hsluoyz — APPROVED** [review]

**Pull request merged**